Fix environment variable masking with flatpak 1.10+

nanonyme opened the pull request:
<!-- If this pull request updates Steam bootstrapper or steamcmd, the relevant commits must contain the updated version number of said components. -->
There's several environment variables which if leaked from host can break apps. Flatpak 1.10 has a behavioural change such that setting variables empty no longer removes them from environment but instead creates empty environment variables. We need to support both old and new flatpak so manifest needs to clear the variables in both old and new way.

Fixes https://github.com/flathub/com.valvesoftware.Steam/issues/729 https://github.com/flathub/com.valvesoftware.Steam/issues/708

gasinvein commented on steam_wrapper/steam_wrapper.py:
```diff
@@ -351,6 +352,11 @@ def configure_shared_library_guard():
+def clear_empty_environment(current_info):
+    environment = current_info["environment"]
+    for key, value in os.environ.items():
+        if environment.get(key.lower()) == "":
```
Why convert to lowercase? I believe environment variables are case-sensitive and e.g. `TZ` and `tz` are actually different variables.

nanonyme replied:
Python configparser quirk. Keys are lowercase.

gasinvein replied:
Ah, right. I think we better fix that instead. Probably set `flatpak_info.optionxform = str` in `read_flatpak_info()` before loading the file would be enough?

LoneFenris commented:
As a note, I see that this only became an issue as of flatpak 1.10. Prior to that, env vars set to the empty string *would* unset them. The flatpak 1.10+ way of unsetting an env var seems to be using the --unset-env arg which writes to a new "unset-environment" entry in the metadata. 
From the documentation, it appears was done in order to allow packages to set empty environment variables.

nanonyme commented:
> As a note, I see that this only became an issue as of flatpak 1.10. Prior to that, env vars set to the empty string _would_ unset them. The flatpak 1.10+ way of unsetting an env var seems to be using the --unset-env arg which writes to a new "unset-environment" entry in the metadata.
> From the documentation, it appears was done in order to allow packages to set empty environment variables.

That should be fine. If the variables are not in your environment, then this algorithm will not do anything.

gasinvein commented:
This change will prevent users from adding their own overrides with env vars set to actual empty strings. But this seems a very unlikely use-case.

nanonyme commented:
> This change will prevent users from adding their own overrides with env vars set to actual empty strings. But this seems a very unlikely use-case.

Right. That's what I was thinking of. One option is changing from empty value to some specific value that signifies that the value should be removed from environment but I think it's a bit hacky.

gasinvein commented on steam_wrapper/steam_wrapper.py:
```diff
@@ -351,6 +353,11 @@ def configure_shared_library_guard():
+    environment = current_info["environment"]
+    for key, value in os.environ.items():
+        if environment.get(key) == "":
+            del os.environ[key]
```
Maybe spit a debug log here just in case?

nanonyme replied:
Added logging, hopefully one that will not confuse the end user.

LoneFenris commented:
> > As a note, I see that this only became an issue as of flatpak 1.10. Prior to that, env vars set to the empty string _would_ unset them. The flatpak 1.10+ way of unsetting an env var seems to be using the --unset-env arg which writes to a new "unset-environment" entry in the metadata.
> > From the documentation, it appears was done in order to allow packages to set empty environment variables.
> 
> That should be fine. If the variables are not in your environment, then this algorithm will not do anything. Taking into a use a new option that won't work on older flatpak than 1.10 is not an option at this point. **I'm still trying to maintain compatibility up to 1.0.** We could have some improvement item to take new parameters into use in a couple of years.

Is that due to the version of flatpak you're building with? If not, and you're building with flatpak 1.10+, you could do both. That is - set the env var to an empty string with --env= *and* unset with --unset-env=. "unset-environment" is set up to take precedence obver env vars set in the Environment section in order to allow proper compatibility.

nanonyme committed:
Support flatpak 1.10
Now unset environment variables are supposed to use --unset-env.
But to support old versions of flatpak we need both stanzas

gasinvein commented on steam_wrapper/steam_wrapper.py:
```diff
@@ -351,6 +353,12 @@ def configure_shared_library_guard():
+    environment = current_info["environment"]
+    for key, value in os.environ.items():
+        if environment.get(key) == "":
+            logging.debug(f"Ensuring {key} not in environment")
```
I'd rather word it "Removing environment variable {key}". I mean, just say what we do, not why we do it.

nanonyme replied:
I'm going to see if Flathub supports --unset-env. It allows doing this thing with no wrapper changes.

nanonyme commented:
> > > As a note, I see that this only became an issue as of flatpak 1.10. Prior to that, env vars set to the empty string _would_ unset them. The flatpak 1.10+ way of unsetting an env var seems to be using the --unset-env arg which writes to a new "unset-environment" entry in the metadata.
> > > From the documentation, it appears was done in order to allow packages to set empty environment variables.
> > 
> > 
> > That should be fine. If the variables are not in your environment, then this algorithm will not do anything. Taking into a use a new option that won't work on older flatpak than 1.10 is not an option at this point. **I'm still trying to maintain compatibility up to 1.0.** We could have some improvement item to take new parameters into use in a couple of years.
> 
> Is that due to the version of flatpak you're building with? If not, and you're building with flatpak 1.10+, you could do both. That is - set the env var to an empty string with --env= _and_ unset with --unset-env=. "unset-environment" is set up to take precedence over env vars set in the Environment section in order to allow proper compatibility. Older versions will ignore "unset-environment" and unset via the empty string. New versions obey "unset-environment".

Right, that worked perfectly. Reflected description and title.

LoneFenris commented:
Agreed about the cluttery-ness. That seems to be the way flatpak recommends doing things for now, though. :shrug:
